pkg/controllers/state/state.go

```diff
@@ -0,0 +1,241 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"knative.dev/pkg/logging"
+
+	"golang.org/x/time/rate"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	podutils "github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Cluster maintains cluster state that is often needed but expensive to compute.
+type Cluster struct {
+	ctx         context.Context
+	rateLimiter *rate.Limiter
+	kubeClient  client.Client
+
+	// Pod Specific Tracking
+	knownPods        sync.Map // mapping of pod namespaced name to struct{} used to track if we've seen a pod before
+	antiAffinityPods sync.Map // mapping of pod namespaced name to *v1.Pod of pods that have required anti affinities
+
+	// Node Status & Pod -> Node Binding
+	mu       sync.RWMutex
```

Comment:
A pattern I see regularly is 
```suggestion
	sync.RWMutex
```
This lets you do nice idiomatic things like `cluster.Lock()`

Reply:
I don't want to expose the Lock/Unlock method here, to me they're not part of the public interface and just an implementation detail.  I like the "mu hat" style where the mutex sits on top of a block of the fields that it proectes:  https://talks.golang.org/2014/readability.slide#21